### PR TITLE
Added Login with Github Button

### PR DESCRIPTION
**Pull request by ryachen01**
Added login with Github button as requested in issue #274

### CODE REVIEW

**Comment by gcampax** [comment]
This is great, thank you for taking time to do this!
Before we can go ahead and merge this work, I will need you to sign the [Contributor License Agreement](https://github.com/stanford-oval/almond-cloud/blob/master/CONTRIBUTING.md#licensing).

**Comment by gcampax** [comment]
This pull request **introduces 1 alert** when merging d2db34d221edce574118ef542bbb562ff86f840b into 6ee906456b34bc0cb470ec2147708ba338d1b09f - [view on LGTM.com](https://lgtm.com/projects/g/stanford-oval/almond-cloud/rev/pr-12d8955c137cea087d85fb1e253482a8796843e6)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ryachen01** [comment]
I sent the agreement and corrected the issue

File: util/passport.js [hunk 1]
```diff
@@ -29,7 +30,9 @@ const EngineManager = require('../almond/enginemanagerclient');
 
 var GOOGLE_CLIENT_ID = '739906609557-o52ck15e1ge7deb8l0e80q92mpua1p55.apps.googleusercontent.com';
 
-const { OAUTH_REDIRECT_ORIGIN, GOOGLE_CLIENT_SECRET } = require('../config');
+var GITHUB_CLIENT_ID = process.env.GITHUB_CLIENT_ID;
```
Comment (gcampax) [inline util/passport.js]: This should probably also be in the configuration file, like the secret.

File: util/passport.js [hunk 2]
```diff
@@ -100,6 +103,65 @@ function associateGoogle(user, accessToken, refreshToken, profile, done) {
...
+        });
+    }).then((user) => {
+        if (!user.newly_created)
+            return user;
+
+        // NOTE: we must start the user here because if we do it earlier we're
+        // still inside the transaction, and the master process (which uses a different
+        // database connection) will not see the new user in the database
+        return EngineManager.get().startUser(user.id).then(() => {
+            // asynchronously inject google-account device
```
Comment (gcampax) [inline util/passport.js]: Comment copy-paste here :)

File: util/passport.js [hunk 3]
```diff
@@ -100,6 +103,65 @@ function associateGoogle(user, accessToken, refreshToken, profile, done) {
...
+            EngineManager.get().getEngine(user.id).then((engine) => {
+                return engine.devices.loadOneDevice({ kind: 'com.github',
+                                                      profileId: profile.id,
+                                                      accessToken: accessToken,
+                                                      refreshToken: refreshToken }, true);
```
Comment (gcampax) [inline util/passport.js]: This is not quite correct.
The goal of this piece of code is to automatically load the Github device in Almond if you login with Github (similar to how we load the Google skill if you login with Google). To do so, we must pass whatever parameters the Github device expects. The device is defined [here](https://github.com/stanford-oval/thingpedia-common-devices/blob/master/com.github/device.js), and you can find its OAuth code if you scroll down to `runOAuth2`

File: routes/user.js [hunk 4]
```diff
@@ -68,6 +68,29 @@ router.get('/oauth2/google/callback', passport.authenticate('google'), (req, res
     }
 });
 
+//oauth login with github
+router.get('/oauth2/github', passport.authenticate('github'));
+router.get('/oauth2/github/callback', passport.authenticate('github'), (req, res, next) => {
```
Comment (gcampax) [inline routes/user.js]: Do you need to pass a list of OAuth scopes here? If you use the access token to configure the Github device, you need the scopes to be the same: user, public_repo, repo, repo:status, gist, notifications

**Comment by gcampax** [comment]
Oh, and juding from [Travis](https://travis-ci.org/stanford-oval/almond-cloud/builds/522356664?utm_source=github_status&utm_medium=notification), it looks like you need to add `passport-github` as a dependency. Use `yarn add passport-github` and then commit both `package.json` and `yarn.lock`.

**Comment by ryachen01** [comment]
I’m not sure if I did the migration file right. Everything else should be fine though.

**Comment by ryachen01** [comment]
@gcampax is there anything more I need to do for this or is it ready to be merged?

**Comment by gcampax** [comment]
No this is good to merge, thanks!

**Pull request merged**